handler/src/main/java/io/netty/handler/ssl/OptionalSslHandler.java

```diff
@@ -0,0 +1,117 @@
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext context, ByteBuf in, List<Object> out) throws Exception {
+        if (in.readableBytes() < SslUtils.SSL_RECORD_HEADER_LENGTH) {
+            return;
+        }
+        if (SslHandler.isEncrypted(in)) {
+            handleSsl(context);
+        } else {
+            handleNonSsl(context);
+        }
+    }
+
+    private void handleSsl(ChannelHandlerContext context) {
+        SslHandler sslHandler = null;
+        try {
+            sslHandler = newSslHandler(context, sslContext);
+            context.pipeline().replace(this, newSslHandlerName(), sslHandler);
+            sslHandler = null;
+        } finally {
+            // Since the SslHandler was not inserted into the pipeline the ownership of the SSLEngine was not
+            // transferred to the SslHandler.
+            if (sslHandler != null) {
+                ReferenceCountUtil.safeRelease(sslHandler.engine());
+            }
+        }
+    }
+
+    private void handleNonSsl(ChannelHandlerContext context) {
+        ChannelHandler handler = newNonSslHandler(context);
+        if (handler != null) {
+            context.pipeline().replace(this, newNonSslHandlerName(), handler);
+        } else {
+            context.pipeline().remove(this);
+        }
+    }
+
+    /**
+     * Optionally specify the SSL handler name, this method may return {@code null}.
+     * @return the name of the SSL handler.
+     */
+    protected String newSslHandlerName() {
+        return null;
+    }
+
+    /**
+     * Override to configure the SslHandler eg. {@link SSLParameters#setEndpointIdentificationAlgorithm(String)}.
+     * The hostname and port is not known by this method so servers may want to override this method and use the
+     * {@link SslContext#newHandler(ByteBufAllocator, String, int)} variant.
+     *
+     * @param context the {@link ChannelHandlerContext} to use.
+     * @param sslContext the {@link SSLContext} to use.
+     * @return the {@link SslHandler} which will replace the {@link OptionalSslHandler} in the pipeline if the
+     * traffic is SSL.
+     */
+    protected SslHandler newSslHandler(ChannelHandlerContext context, SslContext sslContext) {
+        return sslContext.newHandler(context.alloc());
+    }
+
+    /**
+     * Optionally specify the non-SSL handler name, this method may return {@code null}.
+     * @return the name of the non-SSL handler.
+     */
+    protected String newNonSslHandlerName() {
+        return null;
+    }
+
+    /**
+     * Override to configure the ChannelHandler.
+     * @param context the {@link ChannelHandlerContext} to use.
+     * @return the {@link ChannelHandler} which will replace the {@link OptionalSslHandler} in the pipeline
+     * or {@code null} to simply remove the {@link OptionalSslHandler} if the traffic is non-SSL.
+     */
+    protected ChannelHandler newNonSslHandler(ChannelHandlerContext context) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "context". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)